worker/worker.py

```diff
@@ -0,0 +1,244 @@
+                    download_success = True
+            finally:
+                logger.debug('Finished downloading dependency %s/%s', parent_uuid, parent_path)
+                self._dependency_manager.finish_download(
+                    parent_uuid, parent_path, download_success)
+        
+        return dependency_path
+
+    def _store_dependency(self, dependency_path, fileobj, filename):
+        try:
+            if filename.endswith('.tar.gz'):
+                un_tar_directory(fileobj, dependency_path, 'gz')
+            else:
+                with open(dependency_path, 'wb') as f:
+                    shutil.copyfileobj(un_gzip_stream(fileobj), f)
+        except:
+            remove_path(dependency_path)
+            raise
+
+    def remove_dependency(self, parent_uuid, parent_path, uuid):
+        """
+        Unregisters that the run with UUID uuid depends on path parent_path in
+        bundle with UUID parent_uuid. This method is safe to call on
+        dependencies that were never added with add_dependency.
+        """
+        assert(not self.shared_file_system)
+        self._dependency_manager.remove_dependency(parent_uuid, parent_path, uuid)
+
+    def _read(self, socket_id, uuid, path, read_args):
+        run = self._get_run(uuid)
+        if run is None:
+            Run.read_run_missing(self._bundle_service, self, socket_id)
+        else:
+            # Reads may take a long time, so do the read in a separate thread.
+            threading.Thread(target=Run.read,
+                             args=(run, socket_id, path, read_args)).start()
+
+    def _write(self, uuid, subpath, string):
+        run = self._get_run(uuid)
+        if run is not None:
+            run.write(subpath, string)
+
+    def _kill(self, uuid):
+        run = self._get_run(uuid)
+        if run is not None:
+            run.kill('Kill requested')
+
+    def _get_run(self, uuid):
+        with self._runs_lock:
+            return self._runs.get(uuid)
+
+    def finish_run(self, uuid):
+        """
+        Registers that the run with the given UUID has finished.
+        """
+        with self._runs_lock:
+            del self._runs[uuid]
+        if not self.shared_file_system:
+            self._dependency_manager.finish_run(uuid)
+
+    def _checkout(self):
+        try:
+            self._bundle_service.checkout(self.id)
+        except BundleServiceException:
+            traceback.print_exc()
+
+    def _upgrade(self):
+        logger.debug('Upgrading')
+        worker_dir = os.path.dirname(os.path.realpath(__file__))
+
+        while True:
+            try:
+                with closing(self._bundle_service.get_code()) as code:
+                    remove_path(worker_dir)
+                    un_tar_directory(code, worker_dir, 'gz')
+                    break
+            except Exception:
+                traceback.print_exc()
+                time.sleep(1)
+
+        exit(123)
```

Comment:
We can have this system-exec the python script directly to obviate the need for worker.sh.